### DISCUSSION TITLE
Microsoft Office365 IMAP Authentication failed.

### DISCUSSION THREAD

**Discussion by Pisteuon** [jstedfast/MailKit#1777] [Q&A]
I am creating an automated application that runs without user interaction to log in to a Microsoft 365 email account to read emails and move them to another folder.

I have registered the app within Microsoft Entra and believe I have all the settings correct.

I do receive a token back and a TokenEndpoint.

However, I am receiving an "Authentication failed." error.

Below is my C# code.  

I have spent a good bit of time on this but I am just stuck.  I assume that there may be something simple I am missing and would appreciate it if anyone can point me in the right direction.

THANKS!

```csharp
public async Task Connect(string username, string password)
{
    MyIdentityLogger myLogger = new MyIdentityLogger();

    var scopes = new string[] {
        "https://ps.outlook.com/.default"
    };

    IConfidentialClientApplication app;

    app = ConfidentialClientApplicationBuilder.Create(applicationID)
           .WithAuthority($"https://login.microsoftonline.com/{tenantID}/v2.0")
           .WithClientSecret(secretValue)
           .WithLogging(myLogger, true)
           .Build();

    var authToken = await app.AcquireTokenForClient(scopes).ExecuteAsync();
    var oauthx = new SaslMechanismOAuth2(username, authToken.AccessToken);
    using (var client = new ImapClient())
    {
        try
        {
            await client.ConnectAsync("outlook.office365.com", 993, SecureSocketOptions.SslOnConnect);
            await client.AuthenticateAsync(oauthx);

            ReadEmails("Inbox", client);
        }
        catch (AuthenticationException ex)
        {
            Console.WriteLine($"Authentication error: {ex.Message}");
            Console.WriteLine($"Error details: {ex.InnerException?.Message}");
        }

        await client.DisconnectAsync(true);

        Console.ReadLine();
    }
}
```

**Comment by jstedfast** [comment]
Have you granted permissions to your app?

Make sure to follow all of the steps in the "Web services" section: https://github.com/jstedfast/MailKit/blob/master/ExchangeOAuth2.md

**Comment by tinmac** [comment]
Hi @jstedfast 

With respect to granting permissions talk about in the comment above, during `Add permission -> APIs my organization uses` there is no `Office 365 Exchange Online`

I've searched through the other members of `APIs my organization uses` and also `Microsoft Api's` and I cant find any that have IMAP or SMTP permissions.

Do you know if the permissions have been moved or if something has changed?

**Comment by jstedfast** [comment]
I'm not sure, I'd recommend following the Microsoft documentation that my docs link to.

**Reply by tinmac** [reply]
Yeah thanks, sorry for posting this here I know its more of an azure issue.
